core/src/main/java/org/elasticsearch/cluster/action/shard/ShardStateAction.java

```diff
@@ -223,9 +227,24 @@ public ShardFailedClusterStateTaskExecutor(AllocationService allocationService,
         @Override
         public BatchResult<ShardRoutingEntry> execute(ClusterState currentState, List<ShardRoutingEntry> tasks) throws Exception {
             BatchResult.Builder<ShardRoutingEntry> batchResultBuilder = BatchResult.builder();
+            Set<ShardRoutingEntry> nonTrivialTasks = Collections.newSetFromMap(new IdentityHashMap<>());
             List<FailedRerouteAllocation.FailedShard> failedShards = new ArrayList<>(tasks.size());
             for (ShardRoutingEntry task : tasks) {
-                failedShards.add(new FailedRerouteAllocation.FailedShard(task.shardRouting, task.message, task.failure));
+                RoutingNodes.RoutingNodeIterator routingNodeIterator =
+                    currentState.getRoutingNodes().routingNodeIter(task.getShardRouting().currentNodeId());
+                if (routingNodeIterator != null) {
+                    for (ShardRouting maybe : routingNodeIterator) {
+                        if (task.getShardRouting().isSameAllocation(maybe)) {
+                            nonTrivialTasks.add(task);
+                            failedShards.add(new FailedRerouteAllocation.FailedShard(task.shardRouting, task.message, task.failure));
+                            break;
+                        }
+                    }
+                }
+                if (!nonTrivialTasks.contains(task)) {
```

Comment:
@bleskes Hold on, I've thought this through before and I think that duplicates are fine. Check my thinking here?

There are two situations in which a duplicate request can occur.

1. Duplicate requests for a shard that does not exist.
2. Duplicate requests for a shard that does exist.

In situation 1., neither task will ever be added to `nonTrivialTasks` (because the shard doesn't exist). In this case, we want to mark both requests as successful and that's okay.

In situation 2., both tasks will be added to `nonTrivialTasks` which is okay because the allocation service can handle the duplicate failed failed shard entries (it will remove the shard from the routing table on the first entry, and just silently ignore the second). In this case, we want to mark both requests as succeeding or failing together, and that's what happens when the failed shard entries are processed on the allocation service.

Duplicate requests can happen when both the primary and replica attempt to fail the same replica shard.